### PR TITLE
[Raiden Weight Sync 3/7] MoE 128-lane layout and padding for TPU GMM kernels

### PR DESCRIPTION
**Pull request by YixuanWang-99**
## Overview
Part of the stacked Raiden weight-sync enablement PR chain replacing #5089.
**Pairs with Tunix [T5]** (`google/tunix`).

**Stack:**
- [M7] https://github.com/AI-Hypercomputer/maxtext/pull/5166: Cross-repo drift guard
- [M1] https://github.com/AI-Hypercomputer/maxtext/pull/5167: Inhomogeneous layer cycle support in raiden_unscan
- **[M2] AI-Hypercomputer/maxtext (this PR)**: MoE 128-lane layout and padding (pairs with Tunix T5)
- [M3] AI-Hypercomputer/maxtext: Target-free streaming weight conversion
- [M5] AI-Hypercomputer/maxtext: Shared expert gate weight_dtype fix
- [M4] AI-Hypercomputer/maxtext: TrainingEngine Raiden FFI integration (pairs with Tunix T2a)
- [M6] AI-Hypercomputer/maxtext: Rollout cleanup

## Details
- Implements 128-lane interleaving and padding in `convert_utils.py` to match TPU GMM kernel requirements (`gmm_v2.py`).
- Adds `compute_padded_moe_mlp_dim` in `convert_utils.py`.
- Adds unit tests in `tests/post_training/unit/convert_utils_test.py`.

## Verification
- `pytest tests/post_training/unit/convert_utils_test.py` (6/6 passed).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/integration/vllm/convert_utils.py [hunk 1]
```diff
@@ -41,39 +41,101 @@
...
+    return int(tp)
+  config_tp = 0
+  if config is not None:
+    config_tp = int(
+        getattr(config, "rollout_tensor_parallelism", 0)
+        or getattr(getattr(config, "cluster", None), "rollout_tensor_parallelism", 0)
+        or getattr(config, "rollout_mesh_tp", 0)
+        or 0
+    )
+  env_tp = int(
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py]: why do we need to use environment variables instead of the standard config?

File: src/maxtext/integration/vllm/convert_utils.py [hunk 2]
```diff
@@ -41,39 +41,101 @@
...
+
+def resolve_prefuse_moe_weights(
+    config: Any, prefuse_moe_weights: Optional[bool] = None
+) -> bool:
+  """Resolves MoE prefuse flag from override, config, or environment."""
+  if prefuse_moe_weights is not None:
+    return bool(prefuse_moe_weights)
+  if config is not None and getattr(config, "prefuse_moe_weights", None) is not None:
+    return bool(config.prefuse_moe_weights)
+  return os.environ.get("PREFUSE_MOE_WEIGHTS", "0").lower() in ("1", "true", "yes")
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py]: Same here. MaxText has a canonical config. We should not invent an alternative configuration language.

File: src/maxtext/integration/vllm/convert_utils.py [hunk 3]
```diff
@@ -41,39 +41,101 @@
...
+  """Checks if running under Pathways TPU proxy environment."""
+  devices = jax.devices()
+  backend_platform = getattr(devices[0], "platform", "").lower() if devices else ""
+  return (backend_platform == "proxy") or (
+      "proxy" in os.environ.get("JAX_PLATFORMS", "")
+      and bool(os.environ.get("JAX_BACKEND_TARGET"))
+  )
+
+
+def get_host_rss_mb() -> float:
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py]: I don't see how this is related to this PR or to `convert_utils.py`

File: src/maxtext/integration/vllm/convert_utils.py [hunk 4]
```diff
@@ -41,39 +41,101 @@
...
+    return (pages * page_size) / (1024.0 * 1024.0)
+  except Exception:
+    import resource  # pylint: disable=g-import-not-at-top
+    scale = 1024.0 if sys.platform.startswith("linux") else (1024.0 * 1024.0)
+    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / scale
+
+
+def is_verify_weights_enabled() -> bool:
+  """Returns whether weight verification / checksum validation is active."""
+  return os.environ.get("VERIFY_WEIGHTS", "").lower() == "true"
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py]: should this be a config setting in types.py instead of environment var?

File: src/maxtext/integration/vllm/convert_utils.py [hunk 5]
```diff
@@ -160,9 +223,40 @@ class ShapeMismatchError(ValueError):
   """Raised when source and target shapes are incompatible."""
 
 
-def _apply_dtype_cast(val: jax.Array | np.ndarray, tgt_dtype: jnp.dtype, src_key: str) -> jax.Array | np.ndarray:
+def reclaim_host_memory() -> None:
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py]: Again, this does not belong to `convert_utils.py`. I am also not convinced that it is needed at all.

File: src/maxtext/integration/vllm/convert_utils.py [hunk 6]
```diff
@@ -509,8 +599,20 @@ def _pad_and_chunk(arr):
...
+    # [Gate_c0 (128), Up_c0 (128), Gate_c1 (128), Up_c1 (128), ...]
+    num_lanes = target_chunk_size // lane_size
+    shape_lanes = list(p_wi_0.shape)
+    shape_lanes[axis + 1] = num_lanes
+    shape_lanes.insert(axis + 2, lane_size)
+    p_wi_0 = p_wi_0.reshape(shape_lanes)
+    p_wi_1 = p_wi_1.reshape(shape_lanes)
+    combined = jnp.stack([p_wi_0, p_wi_1], axis=axis + 2)
+  else:
+    # Fallback when dimension is not divisible by lane_size:
```
Comment (igorts-git) [inline src/maxtext/integration/vllm/convert_utils.py:562]: Would we get gibberish in this case? Should we raise an exception instead?

File: src/maxtext/integration/vllm/torchax_converter/qwen35_moe.py [hunk 7]
```diff
@@ -19,6 +19,7 @@
 import jax
 import jax.numpy as jnp
 
+from maxtext.integration.vllm.moe_padding import pad_to_tpu_lanes
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/torchax_converter/qwen35_moe.py]: wrong import, pad_to_tpu_lanes is in convert_utils.py not moe_padding.py ?

File: src/maxtext/integration/vllm/convert_utils.py [hunk 8]
```diff
@@ -630,3 +730,67 @@ def _scanned_sharding_from_per_layer(
...
+  orig_apply_dtype_cast = getattr(tunix_utils, "_apply_dtype_cast", None)
+  orig_bulk = getattr(tunix_utils, "_bulk_align_and_unstack", None)
+  orig_unstack = getattr(tunix_utils, "_unstack_scanned_param", None)
+  orig_moe_weights = getattr(tunix_utils, "_MOE_MLP_WEIGHTS", None)
+
+  def _compat_wsc(x, shardings):
+    try:
+      return orig_wsc(x, shardings)
+    except AssertionError:
+      return jax.sharding.reshard(x, shardings)
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/convert_utils.py]: why is this code needed

File: src/maxtext/integration/vllm/convert_utils.py [hunk 9]
```diff
@@ -171,7 +261,8 @@ def _apply_dtype_cast(val: jax.Array | np.ndarray, tgt_dtype: jnp.dtype, src_key
         val.dtype,
         tgt_dtype,
     )
-    return val.astype(tgt_dtype)
+    if hasattr(val, "astype"):
```
Comment (khatwanimohit) [inline src/maxtext/integration/vllm/convert_utils.py]: I think we can remove this hasattr check, agents seem to do this check even when not required

**Review by igorts-git — APPROVED** [review]

**Review by khatwanimohit — APPROVED** [review]

**Review by hengtaoguo — APPROVED** [review]
